ILM can get stuck in the hot phase during attempt-rollover step if rollover is removed from the policy

**Elasticsearch version** (`bin/elasticsearch --version`): 7.10.1 on ESS

**Plugins installed**: n/a

**JVM version** (`java -version`):

```
        "vm_name" : "OpenJDK 64-Bit Server VM",
        "vm_version" : "15.0.1+9",
        "vm_vendor" : "AdoptOpenJDK"
```

**OS version** (`uname -a` if on a Unix-like system): CentOS Linux 8 (Core)

**Description of the problem including expected versus actual behavior**:

Removing the rollover action while an index is in the `check-rollover-ready` step will cause it to get stuck in the hot phase without error in ILM explain. This behavior may or may not be by design, or something ILM was not meant to do, so expected behavior could vary:

1. Allow the hot phase to complete to prevent further writes to the index and move to the next desired phase. Or,
2. Show an error in ILM explain to indicate there is a problem moving to the next step.

**Steps to reproduce**:

 1. Decrease the poll interval and up the logging.

    ```
    PUT /_cluster/settings
    {
      "persistent": {
        "indices.lifecycle.poll_interval": "1s",
        "logger.org.elasticsearch.xpack.ilm": "TRACE"
      }
    }
    ```

 1. Add an index template

    ```
    PUT /_template/devtest
    {
      "index_patterns": [
        "devtest-*"
      ],
      "settings": {
        "index": {
          "lifecycle": {
            "name": "devtest",
            "rollover_alias": "devtest"
          },
          "number_of_shards": "1",
          "number_of_replicas": "0"
        }
      }
    }
    ```

 1. Add policy version 1. Note the rollover action in the hot phase. 

    ```
    PUT /_ilm/policy/devtest
    {
      "policy": {
        "phases": {
          "hot": {
            "min_age": "0ms",
            "actions": {
              "rollover": {
                "max_size": "50mb",
                "max_age": "2m"
              }
            }
          },
          "delete": {
            "min_age": "2m",
            "actions": {
              "delete": {}
            }
          }
        }
      }
    }
    ```

1. Create the index

    ```
    PUT /devtest-000001
    {
      "aliases": {
        "devtest": {
          "is_write_index": true
        }
      }
    }
    ```

1. Let ILM perform an index rollover.

1. Create version 2 of the policy by removing the rollover action from the hot phase. The intention here is to let ILM perform index deletions per policy and prevent creation of a new index.

    ```
    PUT /_ilm/policy/devtest
    {
      "policy": {
        "phases": {
          "hot": {
            "min_age": "0ms",
            "actions": {}
          },
          "delete": {
            "min_age": "2m",
            "actions": {
              "delete": {}
            }
          }
        }
      }
    }
    ```

1. Wait a few minutes and observe the last index is stuck in the rollover action of the hot phase.

    ```
    GET /devtest/_ilm/explain
    {
      "indices" : {
        "devtest-000031" : {
          "index" : "devtest-000031",
          "managed" : true,
          "policy" : "devtest",
          "lifecycle_date_millis" : 1609189822495,
          "age" : "7.12m",
          "phase" : "hot",
          "phase_time_millis" : 1609189822954,
          "action" : "rollover",
          "action_time_millis" : 1609189825089,
          "step" : "check-rollover-ready",
          "step_time_millis" : 1609189825089,
          "phase_execution" : {
            "policy" : "devtest",
            "phase_definition" : {
              "min_age" : "0ms",
              "actions" : {
                "rollover" : {
                  "max_size" : "50mb",
                  "max_age" : "2m"
                }
              }
            },
            "version" : 3,
            "modified_date_in_millis" : 1609186540697
          }
        }
      }
    }
    ```

**Provide logs (if relevant)**:

```
[es/i-0/es.log] [2020-12-28T19:40:46.669Z][TRACE][org.elasticsearch.xpack.ilm.MoveToNextStepUpdateTask] [instance-0000000000] moving [devtest-000003] to next step ({"phase":"hot","action":"rollover","name":"attempt-rollover"})
[es/i-0/es.log] [2020-12-28T19:40:46.669Z][TRACE][org.elasticsearch.xpack.ilm.PolicyStepsRegistry] [instance-0000000000] parsed steps for policy [devtest] in phase [delete], definition: [{"policy":"devtest","phase_definition":{"min_age":"2m","actions":{"delete":{"delete_searchable_snapshot":true}}},"version":2,"modified_date_in_millis":1609184409141}], steps: [[{"phase":"delete","action":"delete","name":"wait-for-shard-history-leases"} => {"phase":"delete","action":"delete","name":"cleanup-snapshot"}, {"phase":"delete","action":"delete","name":"cleanup-snapshot"} => {"phase":"delete","action":"delete","name":"delete"}, {"phase":"delete","action":"delete","name":"delete"} => {"phase":"delete","action":"complete","name":"complete"}, {"phase":"delete","action":"complete","name":"complete"} => null]]
[es/i-0/es.log] [2020-12-28T19:40:46.674Z][ERROR][org.elasticsearch.cluster.service.MasterService] [instance-0000000000] exception thrown by listener notifying of failure from [ilm-move-to-step {policy [devtest], index [devtest-000003], currentStep [{"phase":"hot","action":"rollover","name":"check-rollover-ready"}], nextStep [{"phase":"hot","action":"rollover","name":"attempt-rollover"}]}]
org.elasticsearch.ElasticsearchException: policy [devtest] for index [devtest-000003] failed trying to move from step [{"phase":"hot","action":"rollover","name":"check-rollover-ready"}] to step [{"phase":"hot","action":"rollover","name":"attempt-rollover"}].
	at org.elasticsearch.xpack.ilm.MoveToNextStepUpdateTask.onFailure(MoveToNextStepUpdateTask.java:92) ~[?:?]
	at org.elasticsearch.cluster.service.MasterService$SafeClusterStateTaskListener.onFailure(MasterService.java:513) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.notifyFailedTasks(MasterService.java:446) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:220) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:73) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:151) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [elasticsearch-7.10.1.jar:7.10.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]
	Suppressed: java.lang.IllegalArgumentException: step [{"phase":"hot","action":"rollover","name":"attempt-rollover"}] for index [devtest-000003] with policy [devtest] does not exist
		at org.elasticsearch.xpack.ilm.IndexLifecycleTransition.validateTransition(IndexLifecycleTransition.java:84) ~[?:?]
		at org.elasticsearch.xpack.ilm.IndexLifecycleTransition.moveClusterStateToStep(IndexLifecycleTransition.java:105) ~[?:?]
		at org.elasticsearch.xpack.ilm.MoveToNextStepUpdateTask.execute(MoveToNextStepUpdateTask.java:74) ~[?:?]
		at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:47) ~[elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:702) ~[elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:324) ~[elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:219) [elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:73) [elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:151) [elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678) [elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [elasticsearch-7.10.1.jar:7.10.1]
		at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [elasticsearch-7.10.1.jar:7.10.1]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
		at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: java.lang.IllegalArgumentException: step [{"phase":"hot","action":"rollover","name":"attempt-rollover"}] for index [devtest-000003] with policy [devtest] does not exist
	at org.elasticsearch.xpack.ilm.IndexLifecycleTransition.validateTransition(IndexLifecycleTransition.java:84) ~[?:?]
	at org.elasticsearch.xpack.ilm.IndexLifecycleTransition.moveClusterStateToStep(IndexLifecycleTransition.java:105) ~[?:?]
	at org.elasticsearch.xpack.ilm.MoveToNextStepUpdateTask.execute(MoveToNextStepUpdateTask.java:74) ~[?:?]
	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:47) ~[elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:702) ~[elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:324) ~[elasticsearch-7.10.1.jar:7.10.1]
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:219) ~[elasticsearch-7.10.1.jar:7.10.1]
	... 10 more
```

### Use Case

The issue was discovered trying to use ILM to "dynamically" remove indices once they are no longer needed. The goal is to 1) stop creation of new indices; and 2) let ILM delete indices per the lifecycle policy.

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/PhaseCacheManagement.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/PhaseCacheManagement.java
@@ -224,8 +224,8 @@ public final class PhaseCacheManagement {
      * information, returns null.
      */
     @Nullable
-    static Set<Step.StepKey> readStepKeys(final NamedXContentRegistry xContentRegistry, final Client client,
-                                          final String phaseDef, final String currentPhase, final XPackLicenseState licenseState) {
+    public static Set<Step.StepKey> readStepKeys(final NamedXContentRegistry xContentRegistry, final Client client,
+                                                 final String phaseDef, final String currentPhase, final XPackLicenseState licenseState) {
         final PhaseExecutionInfo phaseExecutionInfo;
         try (XContentParser parser = JsonXContent.jsonXContent.createParser(xContentRegistry,
             DeprecationHandler.THROW_UNSUPPORTED_OPERATION, phaseDef)) {

--- a/x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/ChangePolicyforIndexIT.java
+++ b/x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/ChangePolicyforIndexIT.java
@@ -23,6 +23,7 @@ import org.elasticsearch.xpack.core.ilm.LifecycleSettings;
 import org.elasticsearch.xpack.core.ilm.Phase;
 import org.elasticsearch.xpack.core.ilm.PhaseCompleteStep;
 import org.elasticsearch.xpack.core.ilm.RolloverAction;
+import org.elasticsearch.xpack.core.ilm.SetPriorityAction;
 import org.elasticsearch.xpack.core.ilm.Step.StepKey;
 import org.elasticsearch.xpack.core.ilm.WaitForRolloverReadyStep;
 
@@ -33,6 +34,9 @@ import java.util.concurrent.TimeUnit;
 
 import static java.util.Collections.singletonMap;
 import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.xpack.TimeSeriesRestDriver.createIndexWithSettings;
+import static org.elasticsearch.xpack.TimeSeriesRestDriver.createNewSingletonPolicy;
+import static org.elasticsearch.xpack.TimeSeriesRestDriver.indexDocument;
 
 public class ChangePolicyforIndexIT extends ESRestTestCase {
 
@@ -122,6 +126,37 @@ public class ChangePolicyforIndexIT extends ESRestTestCase {
         Map<String, Object> indexSettings = getIndexSettingsAsMap(indexName);
         String includesAllocation = (String) indexSettings.get("index.routing.allocation.include._name");
         assertEquals("javaRestTest-0,javaRestTest-1,javaRestTest-2,javaRestTest-3", includesAllocation);
+    }
+
+    public void testILMHonoursTheCachedPhaseAfterPolicyUpdate() throws Exception {
+        String indexName = "test-000001";
+        String policyName = "rolloverPolicy";
+        String alias = "thealias";
+        createNewSingletonPolicy(client(), policyName, "hot", new RolloverAction(null, null, null, 1L));
+
+        createIndexWithSettings(client(), indexName, alias, Settings.builder()
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            .put(RolloverAction.LIFECYCLE_ROLLOVER_ALIAS, alias)
+            .put(LifecycleSettings.LIFECYCLE_NAME, policyName));
+
+        // Check the index is on the check-rollover-ready step
+        assertBusy(() -> assertStep(indexName, new StepKey("hot", RolloverAction.NAME, WaitForRolloverReadyStep.NAME)), 30,
+            TimeUnit.SECONDS);
+
+        // update the policy to not contain rollover
+        createNewSingletonPolicy(client(), policyName, "hot", new SetPriorityAction(200));
+
+        // Check the index is on the check-rollover-ready step
+        assertBusy(() -> assertStep(indexName, new StepKey("hot", RolloverAction.NAME, WaitForRolloverReadyStep.NAME)), 30,
+            TimeUnit.SECONDS);
+
+        indexDocument(client(), indexName, true);
+
+        String rolloverIndex = "test-000002";
+        // let's check the cached rollover action still executed and the rollover index exists
+        assertBusy(() -> indexExists(rolloverIndex), 30, TimeUnit.SECONDS);
+        assertBusy(() -> assertStep(indexName, PhaseCompleteStep.finalStep("hot").getKey()), 30, TimeUnit.SECONDS);
     }
 
     private void assertStep(String indexName, StepKey expectedStep) throws IOException {

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java
@@ -43,6 +43,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.Set;
 import java.util.function.BiFunction;
 import java.util.function.LongSupplier;
 
@@ -84,8 +85,13 @@ public final class IndexLifecycleTransition {
                 "], currently: [" + realKey + "]");
         }
 
-        // Always allow moving to the terminal step, even if it doesn't exist in the policy
-        if (stepRegistry.stepExists(indexPolicySetting, newStepKey) == false && newStepKey.equals(TerminalPolicyStep.KEY) == false) {
+        final Set<Step.StepKey> cachedStepKeys =
+            stepRegistry.parseStepKeysFromPhase(lifecycleState.getPhaseDefinition(), lifecycleState.getPhase());
+        boolean isNewStepCached = cachedStepKeys != null && cachedStepKeys.contains(newStepKey);
+
+        // Always allow moving to the terminal step or to a step that's present in the cached phase, even if it doesn't exist in the policy
+        if (isNewStepCached == false &&
+            (stepRegistry.stepExists(indexPolicySetting, newStepKey) == false && newStepKey.equals(TerminalPolicyStep.KEY) == false)) {
             throw new IllegalArgumentException("step [" + newStepKey + "] for index [" + idxMeta.getIndex().getName() +
                 "] with policy [" + indexPolicySetting + "] does not exist");
         }

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java
@@ -14,15 +14,15 @@ import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.Diff;
 import org.elasticsearch.cluster.DiffableUtils;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
-import org.elasticsearch.core.Nullable;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
-import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.common.xcontent.DeprecationHandler;
 import org.elasticsearch.common.xcontent.NamedXContentRegistry;
 import org.elasticsearch.common.xcontent.XContentParseException;
 import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.license.XPackLicenseState;
 import org.elasticsearch.xpack.core.ClientHelper;
@@ -34,6 +34,7 @@ import org.elasticsearch.xpack.core.ilm.LifecyclePolicy;
 import org.elasticsearch.xpack.core.ilm.LifecyclePolicyMetadata;
 import org.elasticsearch.xpack.core.ilm.LifecycleSettings;
 import org.elasticsearch.xpack.core.ilm.Phase;
+import org.elasticsearch.xpack.core.ilm.PhaseCacheManagement;
 import org.elasticsearch.xpack.core.ilm.PhaseExecutionInfo;
 import org.elasticsearch.xpack.core.ilm.Step;
 import org.elasticsearch.xpack.core.ilm.TerminalPolicyStep;
@@ -45,6 +46,7 @@ import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 import java.util.SortedMap;
 import java.util.TreeMap;
 import java.util.stream.Collectors;
@@ -141,6 +143,15 @@ public class PolicyStepsRegistry {
                 }
             }
         }
+    }
+
+    /**
+     * Parses the step keys from the {@code phaseDef} for the given phase.
+     * Returns null if there's a parsing error.
+     */
+    @Nullable
+    public Set<Step.StepKey> parseStepKeysFromPhase(String phaseDef, String currentPhase) {
+        return PhaseCacheManagement.readStepKeys(xContentRegistry, client, phaseDef, currentPhase, licenseState);
     }
 
     private List<Step> parseStepsFromPhase(String policy, String currentPhase, String phaseDef) throws IOException {

--- a/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java
+++ b/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java
@@ -39,8 +39,10 @@ import org.elasticsearch.xpack.core.ilm.OperationMode;
 import org.elasticsearch.xpack.core.ilm.Phase;
 import org.elasticsearch.xpack.core.ilm.PhaseCompleteStep;
 import org.elasticsearch.xpack.core.ilm.RolloverAction;
+import org.elasticsearch.xpack.core.ilm.RolloverStep;
 import org.elasticsearch.xpack.core.ilm.SetPriorityAction;
 import org.elasticsearch.xpack.core.ilm.Step;
+import org.elasticsearch.xpack.core.ilm.WaitForRolloverReadyStep;
 
 import java.io.IOException;
 import java.util.ArrayList;
@@ -489,8 +491,58 @@ public class IndexLifecycleTransitionTests extends ESTestCase {
         try {
             IndexLifecycleTransition.validateTransition(indexMetadata, currentStepKey, nextStepKey, policyRegistry);
         } catch (Exception e) {
-            logger.error(e);
+            logger.error(e.getMessage(), e);
             fail("validateTransition should not throw exception on valid transitions");
+        }
+    }
+
+    public void testValidateTransitionToCachedStepMissingFromPolicy() {
+        LifecycleExecutionState.Builder executionState = LifecycleExecutionState.builder()
+            .setPhase("hot")
+            .setAction("rollover")
+            .setStep("check-rollover-ready")
+            .setPhaseDefinition("{\n" +
+                "        \"policy\" : \"my-policy\",\n" +
+                "        \"phase_definition\" : {\n" +
+                "          \"min_age\" : \"20m\",\n" +
+                "          \"actions\" : {\n" +
+                "            \"rollover\" : {\n" +
+                "              \"max_age\" : \"5s\"\n" +
+                "            },\n" +
+                "            \"set_priority\" : {\n" +
+                "              \"priority\" : 150\n" +
+                "            }\n" +
+                "          }\n" +
+                "        },\n" +
+                "        \"version\" : 1,\n" +
+                "        \"modified_date_in_millis\" : 1578521007076\n" +
+                "      }");
+
+        IndexMetadata meta = buildIndexMetadata("my-policy", executionState);
+
+        Map<String, LifecycleAction> actions = new HashMap<>();
+        actions.put(SetPriorityAction.NAME, new SetPriorityAction(100));
+        Phase hotPhase = new Phase("hot", TimeValue.ZERO, actions);
+        Map<String, Phase> phases = Collections.singletonMap("hot", hotPhase);
+        LifecyclePolicy policyWithoutRollover = new LifecyclePolicy("my-policy", phases);
+        LifecyclePolicyMetadata policyMetadata = new LifecyclePolicyMetadata(policyWithoutRollover, Collections.emptyMap(), 2L, 2L);
+
+        ClusterState existingState = ClusterState.builder(ClusterState.EMPTY_STATE)
+            .metadata(Metadata.builder(Metadata.EMPTY_METADATA)
+                .put(meta, false)
+                .build())
+            .build();
+        try (Client client = new NoOpClient(getTestName())) {
+            Step.StepKey currentStepKey = new Step.StepKey("hot", RolloverAction.NAME, WaitForRolloverReadyStep.NAME);
+            Step.StepKey nextStepKey = new Step.StepKey("hot", RolloverAction.NAME, RolloverStep.NAME);
+            Step currentStep = new WaitForRolloverReadyStep(currentStepKey, nextStepKey, client, null, null, null, 1L);
+            try {
+                IndexLifecycleTransition.validateTransition(meta, currentStepKey, nextStepKey, createOneStepPolicyStepRegistry("my-policy",
+                    currentStep));
+            } catch (Exception e) {
+                logger.error(e.getMessage(), e);
+                fail("validateTransition should not throw exception on valid transitions");
+            }
         }
     }
 

